Moved enterprise permissions to enterprise console (2 of 3)

orangejenny opened the pull request:
## Summary
See https://github.com/dimagi/commcare-hq/pull/29863

This PR updates the enterprise permissions UI and moves it into the Enterprise Console area. It also updates enterprise permissions to use the new `BillingAccount` attributes added in the previous PR instead of the old `DomainPermissionsMirror` config.

![Screen Shot 2021-06-10 at 9 06 54 PM](https://user-images.githubusercontent.com/1486591/121615880-de7deb80-ca2f-11eb-9aa2-a50de9710ad4.png)

This branches off of https://github.com/dimagi/commcare-hq/pull/29863, first new commit is "Moved page from users area to enterprise area"

## Feature Flag
Enterprise Permissions

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

Some, in this PR.

### QA Plan

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-3106)

### Safety story
This is permissions-related, so it's somewhat risky. Impact should be limited to USH.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

orangejenny committed:
Moved page from users area to enterprise area

orangejenny committed:
Moved template

orangejenny committed:
Moved views from users to enterprise app

orangejenny committed:
Updated 'add' form to update source domain

orangejenny committed:
Split mirror domains list into an ignored list and a controlled list

orangejenny committed:
Updated 'delete' form to toggle status of domain

orangejenny committed:
Replaced Subscription code with BillingAccount.get_domains

orangejenny committed:
Added breadcrumbs

orangejenny committed:
Updated UI

orangejenny committed:
Renamed allow_mirroring => allow_enterprise

orangejenny committed:
Renamed get_mirror_domain_links_for_dropdown

orangejenny committed:
Renamed more 'mirror' code

orangejenny committed:
Replaced mirror_domains with get_enterprise_permissions_domains

orangejenny committed:
Updated tests

orangejenny committed:
Moved tests

orangejenny committed:
Updated tests

orangejenny committed:
Switched get_enterprise_permissions_domains to a class method

This is just so the method can incorporate handling cases where an account isn't found.

snopoke commented on corehq/apps/enterprise/views.py:
```diff
@@ -323,3 +329,65 @@ def refresh_item(self, item_id):
+
+@require_superuser
+@require_POST
+def toggle_enterprise_permission(request, domain, target_domain):
```
Rather than toggling I think it it generally safer to pass in the desired value from the view to make sure that the change reflects the user's intent.

orangejenny replied:
a2848c0901f27f23c31e8cf47f7d9be57ca2de98

snopoke commented on corehq/apps/accounting/models.py:
```diff
@@ -500,13 +507,24 @@ def _get_account_by_created_by_domain(cls, domain):
         return StripePaymentMethod.objects.get(web_user=self.auto_pay_user).get_autopay_card(self)
 
+    def get_domains(self):
+        subscriptions = Subscription.visible_objects.filter(account_id=self.id, is_active=True)
+        return [s.subscriber.domain for s in subscriptions]
```
This would be more efficient as a single query:
```suggestion
        return list(Subscription.visible_objects.filter(account_id=self.id, is_active=True).values_list('subscriber__domain', flat=True))
```

snopoke commented on corehq/apps/accounting/models.py:
```diff
@@ -500,13 +507,24 @@ def _get_account_by_created_by_domain(cls, domain):
+    def get_enterprise_permissions_domains(cls, domain):
+        account = cls.get_account_by_domain(domain)
+        if account and account.permissions_source_domain:
+            return list(set(account.get_domains()) - set(account.permissions_ignore_domains))
```
Shouldn't this return an empty list if `domain` is in the ignored list?

orangejenny replied:
resolved by the switch to `EnterprisePermissions`

snopoke commented on corehq/apps/accounting/models.py:
```diff
@@ -500,13 +507,24 @@ def _get_account_by_created_by_domain(cls, domain):
         return BillingAccount.objects.filter(is_customer_billing_account=True, restrict_signup=True)
 
+    @classmethod
+    def get_enterprise_permissions_domains(cls, domain):
```
considering that this is called in places like middleware I think it should be cached.

orangejenny replied:
7ed5564a280d32f0eb3bdee2d2e4c3503f5a5fb4

snopoke commented on corehq/apps/enterprise/views.py:
```diff
@@ -323,3 +329,65 @@ def refresh_item(self, item_id):
+        account.permissions_source_domain = None
+        account.permissions_ignore_domains = []
+    account.save()
+    messages.success(request, _('Permissions saved.'))
```
nit: message could be a bit more explicit

orangejenny replied:
part of 351136851cf787a2834805954a32ec493aaf5433

snopoke commented on corehq/tabs/tabclasses.py:
```diff
@@ -1584,6 +1571,18 @@ def sidebar_items(self):
                 })
+        if self.couch_user.is_superuser:
+            from corehq.apps.accounting.models import BillingAccount
+            if toggles.DOMAIN_PERMISSIONS_MIRROR.enabled_for_request(self._request) \
```
this toggle doesn't seem to be uniformly applied to the code that's impacted by this feature. Some places (like imports) check the toggle but other places don't. Is that something that needs to be addressed (maybe not in this PR)?

orangejenny replied:
Yeah, fair point, but I'm going to leave it for SaaS when they make this generally available, which should include removing the feature flag altogether.

snopoke reviewed (commented):
If you want to turn this off how would you do that? Is there a way to clear the source domain? Is it that tiny 'x' in the select box? Perhaps that needs to be make more obvious?

orangejenny committed:
Fixed some of the failing tests

orangejenny committed:
Merge branch 'jls/move-enterprise-permissions-1-of-3' into jls/move-enterprise-permissions-2-of-3

orangejenny committed:
Replaced BillingAccount.get_enterprise_permissions_domains with EnterprisePermissions methods

orangejenny committed:
Update corehq/apps/accounting/models.py

Co-authored-by: Simon Kelly <skelly@dimagi.com>

orangejenny committed:
Fixed get_domains to only get domains controlled by the domain param

orangejenny committed:
Updated create_enterprise_permissions to use new model

orangejenny committed:
Updated views to use new model

orangejenny committed:
Added docstrings

orangejenny committed:
Added docstrings, removed unnecessary check

orangejenny committed:
Removed arg name in calls

orangejenny committed:
Updated text

orangejenny committed:
Added missing param

orangejenny committed:
Added button for disabling feature

orangejenny committed:
Split toggle view into two views

orangejenny committed:
Removed unnecessary fetch of BillingAccount

orangejenny committed:
Added caching

orangejenny committed:
Fixed feedback messages

orangejenny commented:
@snopoke Thanks for feedback, this is ready for re-review.

> If you want to turn this off how would you do that? Is there a way to clear the source domain? Is it that tiny 'x' in the select box? Perhaps that needs to be make more obvious?

Updated screenshot in PR - there's now a button for disabling the feature, and you can't clear the dropdown anymore.

orangejenny committed:
Added missing account in create_enterprise_permissions

orangejenny committed:
Merge branch 'jls/move-enterprise-permissions-1-of-3' into jls/move-enterprise-permissions-2-of-3

orangejenny committed:
Added merge migration

orangejenny committed:
Fixed test utility

snopoke commented on corehq/apps/enterprise/views.py:
```diff
@@ -323,3 +329,91 @@ def refresh_item(self, item_id):
+    config = EnterprisePermissions.get_by_domain(domain)
+    config.is_enabled = False
+    config.source_domain = None
+    config.save()
```
any reason not to delete the model entirely?

orangejenny replied:
Just caution in case anything goes amiss in the migration. It'll get deleted in https://github.com/dimagi/commcare-hq/pull/29865

snopoke commented on corehq/apps/enterprise/views.py:
```diff
@@ -323,3 +329,91 @@ def refresh_item(self, item_id):
+        messages.error(request, _("Could not add {}}.").format(target_domain))
+        return HttpResponseRedirect(redirect)
+
+    config.domains.append(target_domain)
```
should this make an attempt to prevent adding duplicates?

orangejenny replied:
yeah, might as well - 11aa0a71edc

snopoke commented on corehq/apps/enterprise/views.py:
```diff
@@ -323,3 +329,91 @@ def refresh_item(self, item_id):
+        messages.error(request, _("Could not remove {}.").format(target_domain))
+        return HttpResponseRedirect(redirect)
+
+    config.domains.remove(target_domain)
```
This would error if the domain didn't exist in the list. It's an edge case but I could imagine it happening at some point.

orangejenny replied:
543cd415a2b

snopoke reviewed (approved):
no blocking comments

orangejenny committed:
Fixed create_enterprise_permissions param in test

orangejenny committed:
Fixed cache keys

orangejenny commented on corehq/apps/enterprise/models.py:
```diff
@@ -1,3 +1,54 @@
+            return cls()
+
+    @classmethod
+    @quickcache(['cls.__name__', 'source_domain'], timeout=7 * 24 * 60 * 60)
```
8067450de4b

orangejenny replied:
8067450de4b

orangejenny committed:
Added protection against ValueError if domain isn't present

orangejenny committed:
Updated view to avoid adding duplicates

orangejenny committed:
Moved saves so they only happen when needed

orangejenny committed:
Merge branch 'master' into jls/move-enterprise-permissions-2-of-3

orangejenny committed:
Updated create_enterprise_permissions to create proper subscriptions

orangejenny committed:
Updated test to create active subscriptions

orangejenny committed:
Moved EnterprisePermissions handling in tests into setUp and tearDown

This is to fix test failures due to accounting objects getting deleted out of order
during domain deletion.

orangejenny committed:
Moved remaining create_enterprise_permissions calls into setup methods

orangejenny committed:
Fixed previous commit

orangejenny committed:
Added cache clearing of dropdown links

orangejenny committed:
Moved domain creation into setUpClass to reduce test time

orangejenny committed:
Revert "Moved domain creation into setUpClass to reduce test time"

This reverts commit 647f12c9f9a19f62df35435877972e9e622f7f90.

mjriley committed:
Fixed typo in scheduled reports

orangejenny committed:
Updated navigation to not show if flag is off and feature is disabled

orangejenny committed:
Updated cache clearing to apply to all domains in account

orangejenny committed:
Merge pull request #29937 from dimagi/mjr/fix-move-enterprise-permissions

Fixed typo in scheduled reports

orangejenny committed:
Merge branch 'jls/move-enterprise-permissions-1-of-3' into jls/move-enterprise-permissions-2-of-3

orangejenny committed:
Updated caching to clear based on source domain, not domain being moved

It's the users in the source domain who are affected by the feature.

orangejenny committed:
Added @require_enterprise_admin

orangejenny committed:
Populated account in default EnterprisePermissions

orangejenny committed:
Merge branch 'master' into jls/move-enterprise-permissions-2-of-3

orangejenny commented:
I set up a private release that includes the previous PR, https://github.com/dimagi/commcare-hq/pull/29863/, and ran those migrations on prod.

Verified that for the 3 production projects using this feature, the EnterprisePermissions objects created match the old DomainPermissionsMirror objects. Test projects have already been notified that their configurations will disappear if they're linking domains that aren't in the same BillingAccount.

Also verified that india and swiss have no DomainMirrorPermissions data, as expected.
